### ISSUE TITLE
Create job progress interface

### ISSUE THREAD

**Issue by richard-jones** [CottageLabs/oacwellcome#9]
Create a page which gives the user the following information and options regarding their current job (created in #8 ):

1. A download option for their originally uploaded spreadsheet.
2. What the status of the upload is - has it been processed into a job yet (depending on #13 )
3. What the status of the processing is - just via a status flag on the job or upload
4. A download option which will trigger the creation of a spreadsheet as per #6 and #10

**Comment by Nimphal** [comment]
The part of the interface which triggers the creation of a spreadsheet, what would be the steps? The output_csv function returns a StringIO object. Should I create a csv from that, and if so, should I use standard python csv library code or is there something in MO that is preferred.

**Comment by emanuil-tolev** [comment]
> The part of the interface which triggers the creation of a spreadsheet, what would be the steps?

Can you confirm that the aim of this bit of the functionality is to:
1. take information from the index
2. create a result spreadsheet on-the-fly
3. serve it to the user

**Comment by Nimphal** [comment]
I believe so, yes.

**Comment by emanuil-tolev** [comment]
Your suggestion sounds perfectly reasonable then - I'll check now and get back to you asap if maybe we need to use a UnicodeWriter here instead to make sure we don't introduce uninteresting problems into our lives later.

**Comment by Nimphal** [comment]
I think I might have found a way to calculate % completed. I see that the Record object has an upload position. So there is a record in ES about the current position of the process. I could use that to calculate % if I can get the total number of rows in the file. It seems counting rows of csv-s is not a light process. It could be done during upload, but I'm not sure what performance effect it might have. If it seems too much, I would suggest we don't use the Process part of the progress interface and in stead simply change the text of the button depending on the upload status: if the upload status in completed, the button will say "Download final result", otherwise it will say "Download current stage".

There is also the option of asking the user to submit the number of rows, but that 1. involves extra work in changing the models and 2. won't be very user friendly, particularly if the spreadsheets are big  enough for us not to want to count them, so is probably not worth considering.

**Comment by emanuil-tolev** [comment]
> in stead simply change the text of the button depending on the upload status: if the upload status in completed, the button will say "Download final result", otherwise it will say "Download current stage"

Probably go with that for now, otherwise we need to look at somehow fast-counting all the CSV rows in the file. If we know no newlines are present (except to delineate rows) this could be sped up significantly by conventional file processing methods - Linux has plenty of highly performant commands to do this sort of thing, and the algorithms are surely available in Python libraries.

There is also the slower method of stream-counting - go through each row but throw them away as you go, so you never load more than one row in memory, it just takes a wee while. And then you go through all the rows again to do the actual processing (which could take far longer than the initial count for the progress %). If we think this feature is worth another loop over the incoming spreadsheet (which should be pretty damn fast in reality), then this is the simplest, most understandable solution for me.

**Comment by emanuil-tolev** [comment]
> Your suggestion sounds perfectly reasonable then - I'll check now and get back to you asap if maybe we need to use a UnicodeWriter here instead to make sure we don't introduce uninteresting problems into our lives later.

Sorry, I had misunderstood the problem here. @richard-jones wrote the output_csv function which outputs a CSV as a string (a StringIO in fact). You simply need to give this to the Flask send_file function, set the as_attachment=True and attachment_filename=lalala.csv parameters.

**Comment by Nimphal** [comment]
I've made pull request, the interface should be completed.

**Comment by emanuil-tolev** [comment]
So to test this fully we'll need the processing bit in as well, otherwise we won't necessarily know the "download sheet in progress" bit works.

**Comment by richard-jones** [comment]
Coming into this conversation a bit late, sorry.

Yes, the output_csv function will give you back as a string the raw
contents of the csv, so you can serve it to the user.  Just looked at what
you've already done, and yes that should be it.


On 8 January 2015 at 19:31, Emanuil Tolev <notifications@github.com> wrote:

> So to test this fully we'll need the processing bit in as well, otherwise
> we won't necessarily know the "download sheet in progress" bit works.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CottageLabs/oacwellcome/issues/9#issuecomment-69234778>
> .
>



-- 

Richard Jones,

Founder, Cottage Labs
t: @richard_d_jones, @cottagelabs
w: http://cottagelabs.com

**Comment by richard-jones** [comment]
I've now added a property to the SpreadsheetJob object which will give you the % completeness of the job

    job.pc_complete

This will return a floating point between 0 and 100, which you can customise for display in the UI (probably to an int, but depends on what looks best).

**Comment by richard-jones** [comment]
Target: by Wednesday (for the meeting) we should be able to upload a file, and have the asynchronous back-end process the job and give us a progress update.

**Comment by richard-jones** [comment]
@Nimphal - if you were feeling adventurous, it would be quite fun to make the % complete auto-update on the page.

If you made an API endpoint like 

    /progress/<job id>/pc

that just returned the number (possibly wrapped in a jsonp response of some kind),

then had a job on the page that used setTimeout to run a call to that every few seconds, you could update it in the page without reloading.  Something like

```javascript
function updatePc() {
    pc = pcAjaxApiCall();
    $("#pc_container").html(pc);
    setTimeout(updatePc, 5000);
}
```

Up to you if you want to have a go at it, though.

**Comment by emanuil-tolev** [comment]
Now that would be cool!

**Comment by richard-jones** [comment]
Actually, I am considering whether we should go further.  It would be great if this page updated itself automatically from "submitted" to "processing" to "complete", and also gave a hint to the user as to when their job will be processed by telling them how many are in front of them in the queue.

Probably this is outside the scope of what you've estimated for, @Nimphal , but we might have a go later at building an API endpoint for the jobs, which returns something like

```json
{
    "job" : { <job record> },
    "pc" : "pc complete",
    "queue" : "number in front of you"
}
```

Then building the UI dynamically from that.

**Comment by Nimphal** [comment]
I am glad we're having a brain storming session about the interface. All of
these ideas are really awesome and cool (the moment the JS pandora box
opens, there's no end), but I doubt I'd be able to implement all of them by
the deadline. I'll look into the percentage thing, since that seems fairly
straightforward.

2015-01-15 11:32 GMT+00:00 Richard Jones <notifications@github.com>:

**Comment by richard-jones** [comment]
Yes, absolutely - I'm thinking out loud about a future version of the page.  If you just get the UI to the state you are happy with design and information wise I'll consider taking it forward at some point.

**Comment by Nimphal** [comment]
Just made a pull request, I am now happy with the buttons and the progress percentage changes dynamically. Will assign to @richard-jones once the PR is merged.

**Comment by emanuil-tolev** [comment]
I've taken this out of the Full Service milestone, it's already beyond what's necessary for the full service :).

@richard-jones note that at least 1 page refresh is needed to see the latest AJAX addition working, because otherwise the spreadsheet is in "submitted" state and doesn't show a % completion. But soon as you refresh the page and the job *has* been picked up, it does increment the %. So it's more useful for long-running jobs than making the progress interface just generally swanky.

**Comment by richard-jones** [comment]
Ok, I've opened a new issue for a fully dynamic progress page, which I'll look at next week if the time presents itself.
